jax/experimental/maps.py

```diff
@@ -1126,122 +1127,85 @@ def restore_units(is_unit, vals):
-        ]
-      if not const_axes_s:  # NOTE: This can be called multiple times
-        const_axes_s.store(const_axes)
-      assert const_axes_s.val == const_axes
-      return (*out_axes_no_units, *const_axes)
-    pe_params = dict(
-        params,
-        in_axes=tuple(a for a, u in zip(in_axes, in_units) if not u),
-        donated_invars=tuple(a for a, u in zip(donated_invars, in_units) if not u),
-        out_axes_thunk=new_out_axes_thunk)
-    outs_no_units = primitive.bind(f, *args_no_units, **pe_params)
-    new_out_axes_thunk()  # Make sure it is called at least once to compute const_axes
-    return restore_units(out_units(), outs_no_units)
-
-  jaxpr, out_pvals, consts, env_tracers = self.partial_eval(
-      f, in_pvals, app, instantiate=False)
-
-  out_axes = params['out_axes_thunk']()
-  const_axes = const_axes_s.val
+  out_axes_thunk = params['out_axes_thunk']
+
+  # Adjust input tracers' pvals for mapped axes, and unpack.
+  in_pvals = [t.pval if t.pval.is_known() else
+              pe.PartialVal.unknown(
+                  _delete_aval_axes(t.pval.get_aval(), axes, global_axis_sizes))
+              for t, axes in zip(tracers, in_axes)]
+  in_knowns, in_avals, in_consts = pe.partition_pvals(in_pvals)
+
+  # Wrap f to perform partial evaluation, and plumb out aux data.
+  f = pe.trace_to_subjaxpr_nounits(f, self.main, False)
+  f, aux = pe.partial_eval_wrapper_nounits(f, tuple(in_knowns), tuple(in_avals))
+  # Also grab the local named shapes of the output (known and res).
+  f, out_named_shapes = out_local_named_shapes(f, frozenset(global_axis_sizes))
+
+  # Adjust params for knowns (donated_invars, in_axes, out_axes_thunk).
+  @as_hashable_function(closure=out_axes_thunk)
+  def new_out_axes_thunk():
+    out_knowns, _, _, _ = aux()
+    _, out_axes = partition_list(out_knowns, out_axes_thunk())
+    return (*out_axes, *res_axes())
+  def res_axes():
+    _, _, jaxpr, _ = aux()
+    num_res = len(jaxpr.constvars)
+    res_named_shapes = out_named_shapes()[-num_res:] if num_res else []
+    sorted_named_shapes = [sorted(ns, key=str) for ns in res_named_shapes]
+    return [AxisNamePos(zip(named_shape, range(len(named_shape))),
+                        user_repr=f'<internal: {named_shape}>')
+            for named_shape in sorted_named_shapes]
+  known_params = dict(
+      params, in_axes=tuple(a for a, k in zip(in_axes, in_knowns) if k),
+      donated_invars=tuple(d for d, k in zip(donated_invars, in_knowns) if k),
+      out_axes_thunk=new_out_axes_thunk)
+
+  # Run the known part.
+  outs = primitive.bind(f, *in_consts, **known_params)
+  out_knowns, out_avals, jaxpr, env = aux()
+  known_outvals, res = split_list(outs, [len(outs) - len(jaxpr.constvars)])
+  with core.extend_axis_env_nd(global_axis_sizes.items()):
+    jaxpr = pe.convert_constvars_jaxpr(jaxpr)
+
+  # Set up new params.
+  out_axes = [a for a, k in zip(out_axes_thunk(), out_knowns) if not k]
```

Comment:
Hrm, what original `out_axes` do you mean? I think at this point in the function logic we only have `out_axes_thunk` and not any `out_axes`.

It's true that `out_axes_thunk` has already been called once at this point, in `new_out_axes_thunk`. So I'll just save it from there.

I didn't realize `out_axes_thunk` could be expensive! Memoization seems reasonable if it doesn't depend on any global state. But I won't do that now since it's a bigger change.